### PR TITLE
Load plugin immediately

### PR DESCRIPTION
**Pull request by kou**
## What was the end-user or developer problem that led to this PR?

We can install RubyGems plugin by "gem install XXX". The installed plugin is used from the NEXT "gem ...".

For example, "gem install gem-src kaminari" doesn't use gem-src plugin for kaminari. "gem install gem-src && gem install kaminari" uses gem-src plugin for kaminari.

## What is your fix for the problem, implemented in this PR?

How about loading a plugin immediately when the plugin is installed? If this proposal is implemented, "gem install gem-src kaminari" works like "gem install gem-src && gem install kaminari".

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by kou** [comment]
Rebased.

**Comment by deivid-rodriguez** [comment]
If I have version X of plugin `gem-src` installed, then that will be loaded right before `gem install gem-src:X+1 kaminari` runs, but also version `X+1` will be loaded right after `gem-src` is installed, correct? Won't that cause two different versions of the same gem to be loaded? That sounds not good.

**Comment by kou** [comment]
> If I have version X of plugin `gem-src` installed, then that will be loaded right before `gem install gem-src:X+1 kaminari` runs, but also version `X+1` will be loaded right after `gem-src` is installed, correct?

Correct.

> Won't that cause two different versions of the same gem to be loaded?

Yes. Two different versions of the same gem are loaded in the situation.

> That sounds not good.

I agree with you.

How about preventing loading `gem-src:X+1` in the situation? `gem install gem-src:X+1 kaminari` loads `gem-src:X` before it's executed and `gem-src:X+1` isn't loaded while `gem install gem-src:X+1 kaminari` is executing.
`gem-src:X+1` is used (and `gem-src:X` isn't used) from the next `gem install ...`.

Is this behavior acceptable?

**Comment by kou** [comment]
I've implemented the behavior.

What do you think about this?

**Comment by deivid-rodriguez** [comment]
@kou I'm fine with the updated behavior, thanks. I will review the PR in a few days, just to give other maintainers time to leave their feedback or raise their objections (if any).

**Comment by kou** [comment]
Thanks!

File: lib/rubygems/installer.rb
Comment (deivid-rodriguez) [inline lib/rubygems/installer.rb:348]: If `run_post_install_hooks` happened before `load_plugin`, then any post install hooks potentially registered by the gem would be run immediately too, right?

For example, say the `rdoc` gem installs a plugin that generates documentation after gems are installed, then `gem install rdoc` would immediately generate docs for `rdoc` itself.

Did you consider this? Do you think it would be desirable?

**Review by deivid-rodriguez — COMMENTED** [review]
There seem to be some error output when running tests now, can you have a look? Could those point to some bad side effect on real usage?

```
/home/runner/work/rubygems/rubygems/tmp/test_rubygems_20230516-2057-9o50ny/gemhome/gems/a-2/lib/rubygems_plugin.rb
Error loading RubyGems plugin "/home/runner/work/rubygems/rubygems/tmp/test_rubygems_20230516-2057-gehefu/gemhome/plugins/a_plugin.rb": cannot load such file --  (LoadError)
Error loading RubyGems plugin "/home/runner/work/rubygems/rubygems/tmp/test_rubygems_20230516-2057-oovia/gemhome/plugins/a_plugin.rb": cannot load such file --  (LoadError)
Error loading RubyGems plugin "/home/runner/work/rubygems/rubygems/tmp/test_rubygems_20230516-2057-rpqs7e/gemhome/plugins/a_plugin.rb": cannot load such file --  (LoadError)
```

File: lib/rubygems/installer.rb
Comment (kou) [reply lib/rubygems/installer.rb:348]: Ah, sorry. I didn't consider this.

`rdoc` (and `yard`) use `done_installing` hook not `post_install` hook but I can understand what you want to say.

I'll move `load_plugin` BEFORE `run_post_install_hooks`.

Comment (kou) [reply lib/rubygems/installer.rb:346]: Done.

BTW, I think that we can remove https://github.com/rubygems/rubygems/blob/master/lib/rubygems/rdoc.rb and and implement this by RubyGems plugin in RDoc after we merge this pull request.

**Comment by kou** [comment]
> There seem to be some error output when running tests now, can you have a look? Could those point to some bad side effect on real usage?
> 
> ```
> /home/runner/work/rubygems/rubygems/tmp/test_rubygems_20230516-2057-9o50ny/gemhome/gems/a-2/lib/rubygems_plugin.rb
> Error loading RubyGems plugin "/home/runner/work/rubygems/rubygems/tmp/test_rubygems_20230516-2057-gehefu/gemhome/plugins/a_plugin.rb": cannot load such file --  (LoadError)
> Error loading RubyGems plugin "/home/runner/work/rubygems/rubygems/tmp/test_rubygems_20230516-2057-oovia/gemhome/plugins/a_plugin.rb": cannot load such file --  (LoadError)
> Error loading RubyGems plugin "/home/runner/work/rubygems/rubygems/tmp/test_rubygems_20230516-2057-rpqs7e/gemhome/plugins/a_plugin.rb": cannot load such file --  (LoadError)
> ```

Sorry. I missed them.
I've fixed them.

**Review by deivid-rodriguez — APPROVED** [review]
Thank you! This looks good to me, let's hope it doesn't create any bad side effects. But it something unexpected comes up, we can consider it later.

File: lib/rubygems/installer.rb
Comment (kou) [reply lib/rubygems/installer.rb:346]: Oh, I didn't notice #6021.
I'll work on it after we merge this.

**Comment by kou** [comment]
Please mention me when we found any unexpected side effect.

File: lib/rubygems/installer.rb
Comment (kou) [reply lib/rubygems/installer.rb:346]: Implementation: https://github.com/ruby/rdoc/pull/1171

**Pull request merged**
